File: front/web/lib/ideas.ts
/**
 * Ideas lifecycle — the Workbench /ideas source (AIDOS step S27).
 *
 * KRD §115/§116/§118/§119: an Idea is a CANDIDATE-truth staged ABOVE product but BELOW the freeze —
 * a sketched body (proposes + intent + provenance) with NO version-freeze and NO mirror. That double
 * absence is EXACTLY what makes it an idea and not a truth. The lifecycle runs
 * draft → grilled → {spiking → harvested | harvested}, with a parallel rejected lane (traced, kept).
 * "Promote" is NOT a status — it is the ACT of writing the idea's mirror, which IS the /goal, which
 * IS the freeze into /kernel. An idea with no mirror can NEVER enter the kernel: Promote without a
 * mirror returns the NO_MIRROR_NO_KERNEL BlockReason (the wall, KRD §44.5).
 *
 * This module is the DECLARED TWIN of the Go package back/kernel/ideas — the SAME five statuses, the
 * SAME §75 gestures, the SAME promotion gate (no mirror ⇒ NO_MIRROR_NO_KERNEL with the fix path
 * write_mirror_run_goal_freeze). One semantics, no drift — so /ideas renders EXACTLY the lifecycle
 * the Go deciders compute. The reproducibility mirror lib/ideas.test.ts (fast-check) pins it.
 *
 * READ-ONLY (CLAUDE.md §7 ui-completeness, the wall): /ideas PROJECTS the lifecycle and the gate
 * verdict; capturing/advancing an idea rides the idea-intake MCP (ideas schema, ABOVE the wall), and
 * a PROMOTION writes a kernel truth via the S20 ChangeSet path under the aidos writer role — never a
 * write from this screen. The agent DB role is SELECT-only on kernel/mirrors (the wall holds).
 */

/** The five lifecycle statuses — the closed set KRD §118 names (no sixth status). */
export type Status = "draft" | "grilled" | "spiking" | "harvested" | "rejected";

/** The layer/kind an idea would become if promoted (KRD §118). */
export type Proposes =
	| "control"
	| "policy"
	| "operation"
	| "action"
	| "entity"
	| "product";

/** The provenance source — the two legal on-ramps (KRD §117). */
export type ProvenanceSource = "human" | "incident";

/** Who/what engendered an idea (KRD §119): source + verbatim detail. */
export interface Provenance {
	source: ProvenanceSource;
	/** The human utterance ("finalement je veux une remise") or the incident ref ("#1234"). */
	detail: string;
}

/**
 * An Idea — a candidate-truth (mirrors ideas.Idea). It carries NO `version` and NO `mirror` field:
 * those two absences are what distinguish an idea from a truth, and the type makes them
 * unrepresentable (there is simply no field for them). The id is the content hash of the sketch.
 */
export interface Idea {
	id: string;
	proposes: Proposes;
	intent: string;
	provenance: Provenance;
	status: Status;
	/** The traced reason an idea was rejected (kept, never deleted). Empty unless rejected. */
	rejectReason?: string;
}

/** The actionable refusal shape (mirrors blockreason.BlockReason, KRD §44.5). */
export interface BlockReason {
	code: string;
	severity: string;
	explanation: string;
	howToFix: string[];
}

/** The legal outcome of promoting an idea: a kernel truth via /goal, back-linked to the idea. */
export interface Promotion {
	ideaId: string;
	mirrorRef: string;
	/** The back-link the new kernel truth carries — the idea it came from (KRD §119). */
	provenanceIdeaId: string;
}

/** The canonical NO_MIRROR_NO_KERNEL BlockReason — the twin of blockreason.For (FR prose). */
export const NO_MIRROR_NO_KERNEL: BlockReason = {
	code: "NO_MIRROR_NO_KERNEL",
	severity: "blocking",
	explanation:
		"Refus du mur : on promeut une idée vers le noyau SANS miroir. Une idée est une vérité-candidate sans gel ni miroir ; la seule porte vers /kernel est idea → miroir → /goal → gel. Sans miroir, aucune idée n'entre jamais dans le noyau.",
	howToFix: [
		"write_mirror_run_goal_freeze : écrivez le miroir BDD rouge de l'idée — ce rouge EST le /goal, et le /goal EST le gel dans /kernel.",
		"assign_authority : faites approuver le /goal par l'autorité du sous-graphe.",
		"rerun aidos check : la promotion passe le mur dès que l'idée porte son miroir.",
	],
};

/** The legal next-gestures from each status (the state machine; mirrors the Go transitions). */
const TRANSITIONS: Record<Status, Status[]> = {
	draft: ["grilled", "rejected"],
	grilled: ["spiking", "harvested", "rejected"],
	spiking: ["harvested", "rejected"],
	harvested: ["rejected"], // promotion is NOT a status — it is the /goal act.
	rejected: [],
};

/** canPromote: only a harvested idea is eligible (and still needs a mirror). */
export function canPromote(idea: Idea): boolean {
	return idea.status === "harvested";
}

/**
 * promote is the GATE — the deterministic twin of ideas.Promote (KRD §116/§44.5). It refuses unless
 * the idea is harvested AND a non-empty mirror reference is supplied. No mirror ⇒ NO_MIRROR_NO_KERNEL
 * and NO kernel write (the idea stays harvested). A non-empty mirror ⇒ a Promotion whose provenance
 * points back to the idea (the freeze itself is the aidos CLI role via /goal, never here). Exactly
 * one of {promotion, block} is set.
 */
export function promote(
	idea: Idea,
	mirrorRef: string,
): { promotion?: Promotion; block?: BlockReason } {
	if (idea.status !== "harvested" || mirrorRef === "") {
		return { block: NO_MIRROR_NO_KERNEL };
	}
	return {
		promotion: {
			ideaId: idea.id,
			mirrorRef,
			provenanceIdeaId: idea.id,
		},
	};
}

/** An idea has no mirror, by definition — the "no mirror yet" marker is always true here. */
export function hasMirror(): boolean {
	return false;
}

/** legalNext returns the gestures legal from an idea's current status (for the board controls). */
export function legalNext(idea: Idea): Status[] {
	return TRANSITIONS[idea.status];
}

/** byLane groups ideas by status into the board's lanes, in canonical lane order. */
export function lanes(): Status[] {
	return ["draft", "grilled", "spiking", "harvested", "rejected"];
}

File: back/kernel/ideas/ideas.go
// Package ideas defines the Idea record and its lifecycle — the ONLY door into
// the Kernel (AIDOS step S27).
//
// An Idea is a CANDIDATE-truth staged ABOVE product but BELOW the freeze (KRD
// §115/§118): it has the form of a truth — a sketched body (`proposes` + `intent`
// + `provenance`) — but it carries NEITHER a version-freeze NOR a mirror. That
// double absence is EXACTLY what makes it an idea and not a truth (KRD §118: "PAS
// de version-gel, PAS de mirror — c'est ce qui la distingue d'une vérité"). The
// Idea type makes both unrepresentable: there is no Version and no Mirror field on
// the struct — by construction an idea cannot carry the thing that would make it a
// truth.
//
// The lifecycle is wired on the §75 gestures (KRD §118):
//
//	          ┌──────────────────── reject (traced) ──────────────────┐
//	          ▼                                                        │
//	draft ──grill──▶ grilled ──spike──▶ spiking ──harvest──▶ harvested ─┴─▶ promote
//	                    └────────────── harvest ──────────────▶ harvested
//
// "Promote" is NOT a status — it is the ACT of writing the idea's mirror, which
// IS the /goal, which IS the freeze into /kernel (KRD §116: "promouvoir une idée
// = écrire son miroir + la geler"). Promote of a harvested idea WITHOUT a mirror
// returns the actionable NO_MIRROR_NO_KERNEL BlockReason (KRD §44.5): an idea with
// no mirror can NEVER enter the kernel. This package proves the GATE and the
// LIFECYCLE; the downstream kernel write is the aidos CLI role via the /goal flow,
// not the agent (the wall, CLAUDE.md §2).
//
// DETERMINISM-FIRST (CLAUDE.md §6/§8): every function here is PURE and TOTAL — no
// clock (timestamps are passed in by the caller / DB), no rng, no I/O, never
// panics. The same input always yields the same output; the reproducibility mirror
// ideas_property_test.go pins that. The id is the content hash of the canonical
// body, REUSING the S01/S02 content-hash scheme (records.Hash/Canonicalize) — never
// a forked hashing path.
package ideas

import (
	"encoding/json"
	"fmt"

	"github.com/steph-frtech/aidos/back/kernel/records"
)

// Status is the lifecycle state of an Idea. The set is CLOSED and small (KRD §118)
// — there is no sixth status, and "promote" is deliberately NOT a status (it is the
// act of writing the mirror = /goal = freeze, modelled by Promote()).
type Status string

const (
	// StatusDraft — just captured from a human utterance or an incident; not yet
	// grilled. The entry state.
	StatusDraft Status = "draft"
	// StatusGrilled — challenged against the domain model (/grill); the intention
	// is sharpened but still not falsifiable.
	StatusGrilled Status = "grilled"
	// StatusSpiking — the "floue ?" branch: exploration with the ratchet OFF
	// (/spike). A spiking idea is being probed, not graduated.
	StatusSpiking Status = "spiking"
	// StatusHarvested — the discovered truth has been extracted (/harvest); the
	// idea is now ready to be promoted — but only by acquiring a mirror.
	StatusHarvested Status = "harvested"
	// StatusRejected — a bad idea, traced and kept (append-only); never deleted.
	StatusRejected Status = "rejected"
)

// Statuses returns the five lifecycle statuses in canonical order. Used by the
// validator and the Workbench projection so the set of lanes is never invented.
func Statuses() []Status {
	return []Status{StatusDraft, StatusGrilled, StatusSpiking, StatusHarvested, StatusRejected}
}

// Proposes is the targeted layer/kind an idea WOULD become if promoted (KRD §118).
// The set is the closed set of kernel source kinds plus "product"; an idea names
// what it proposes, it does not yet freeze it.
type Proposes string

const (
	ProposesControl   Proposes = "control"
	ProposesPolicy    Proposes = "policy"
	ProposesOperation Proposes = "operation"
	ProposesAction    Proposes = "action"
	ProposesEntity    Proposes = "entity"
	ProposesProduct   Proposes = "product"
)

// ProposesKinds returns the closed set of layer/kinds an idea may propose, in
// canonical order. The Workbench renders exactly these; none is invented.
func ProposesKinds() []Proposes {
	return []Proposes{
		ProposesControl, ProposesPolicy, ProposesOperation,
		ProposesAction, ProposesEntity, ProposesProduct,
	}
}

// ProvenanceSource is who/what engendered an idea (KRD §117): a human utterance or
// a reality/incident signal. The set is closed — the two legal on-ramps.
type ProvenanceSource string

const (
	// ProvenanceHuman — "finalement je veux que…": a human intention (KRD §117).
	ProvenanceHuman ProvenanceSource = "human"
	// ProvenanceIncident — reality diverged: an incident #NNNN (KRD §117).
	ProvenanceIncident ProvenanceSource = "incident"
)

// Provenance records who wanted what, when, why (KRD §119): the source and the
// verbatim detail (the human utterance, or the incident reference `#NNNN`). Every
// frozen truth later points back to the idea — and through it to this provenance.
type Provenance struct {
	Source ProvenanceSource `json:"source"`
	// Detail is the human utterance ("finalement je veux une remise") or the
	// incident reference ("#1234"). Verbatim, never paraphrased.
	Detail string `json:"detail"`
}

// Idea is a candidate-truth (KRD §118). It carries a sketched body — Proposes +
// Intent + Provenance — and a lifecycle Status. It carries NO Version (no freeze)
// and NO Mirror: those two absences are what distinguish an idea from a truth, and
// the type makes them UNREPRESENTABLE (there is simply no field for them). The ID
// is the content hash of the canonical body (content-addressing), reusing the
// S01/S02 scheme — so the same sketched body always lands at the same address.
//
// The body that the ID hashes over is {proposes, intent, provenance} — the durable
// identity of the candidate. Status is lifecycle metadata that ADVANCES (it is not
// part of the content address): grilling or harvesting the same sketch does not
// change its identity. RejectReason is set only when Status == rejected (traced).
type Idea struct {
	ID         string     `json:"id"`
	Proposes   Proposes   `json:"proposes"`
	Intent     string     `json:"intent"`
	Provenance Provenance `json:"provenance"`
	Status     Status     `json:"status"`
	// RejectReason is the traced reason an idea was rejected (KRD §118: reject is
	// append-only/traced, never a deletion). Empty unless Status == rejected.
	RejectReason string `json:"reject_reason,omitempty"`
}

// body is the content-addressed JSONB shape: the durable sketch {proposes, intent,
// provenance}. Status and reject_reason are deliberately EXCLUDED from the address
// (lifecycle metadata, not identity) — and there is no `version` and no `mirror`
// key, by construction.
type body struct {
	Proposes   Proposes   `json:"proposes"`
	Intent     string     `json:"intent"`
	Provenance Provenance `json:"provenance"`
}

// ErrParse is returned when an idea body JSONB cannot be decoded into an Idea.
var ErrParse = fmt.Errorf("ideas: malformed idea body")

// Canonicalize returns the deterministic JSONB body an idea is hashed over: the
// sketch {proposes, intent, provenance} with object keys sorted lexicographically.
// It REUSES records.Canonicalize (the S01/S02 content-hash scheme) — never a
// forked hashing path. So records.Hash(Canonicalize(i)) is the idea id.
func Canonicalize(i Idea) ([]byte, error) {
	raw, err := json.Marshal(body{
		Proposes:   i.Proposes,
		Intent:     i.Intent,
		Provenance: i.Provenance,
	})
	if err != nil {
		return nil, fmt.Errorf("%w: %v", ErrParse, err)
	}
	return records.Canonicalize(raw)
}

// Hashed returns i with its ID set to the content hash of its canonical body,
// reusing records.Hash (the same address space as the content store and every
// record kind — never a re-invented hashing path). It is a pure function: the
// same sketch always yields the same id, regardless of input key order.
func Hashed(i Idea) (Idea, error) {
	b, err := Canonicalize(i)
	if err != nil {
		return Idea{}, err
	}
	i.ID = records.Hash(b)
	return i, nil
}
